### PR TITLE
fix: record local alias for dynamic import() destructuring renames

### LINKED ISSUE #1824: Dynamic import destructuring rename (const { a: b } = await import(...)) records the wrong binding name
## Summary

While fixing #1730 (renamed static `import { X as Y }` specifiers not resolving to the correct call edge), I found the same class of bug in the *dynamic* import extraction path: `const { a: b } = await import('./mod')` records the original name (`a`) instead of the local binding actually used at call sites (`b`), so a subsequent call to `b(...)` will not resolve against the exported symbol `a` in `./mod`.

## Root cause

`extractDynamicImportNames` in `src/extractors/javascript.ts` (`pair_pattern` branch, handling `{ a: b } = await import(...)`) currently does:

```js
} else if (child.type === 'pair_pattern') {
  // { a: localName } → use localName (the alias) for the local binding,
  // but use the key (original name) for import resolution
  const key = child.childForFieldName('key');
  if (key) names.push(key.text);
}
```

The comment describes wanting *both* pieces of information, but only `key.text` (`a`, the original exported name) is ever pushed into the flat `names` array — the local alias (`b`) is dropped entirely. Downstream, `buildImportedNamesMap`/`buildImportedNamesForNative` use `imp.names` entries as the *local* binding name to key the `importedNames` map against call-site text (this is exactly the invariant #1730 restored for static imports). Since `names` holds `a` instead of `b` here, a call site `b(...)` never matches this import's map entry.

I have not checked whether the Rust extractor's `extract_dynamic_import_names` (`crates/codegraph-core/src/extractors/javascript.rs`) has the identical issue — likely, given the mirrored-engine convention, but not yet confirmed.

## Repro

```js
async function run() {
  const { realName: alias } = await import('./mod.js');
  alias();
}
```

`alias()` should resolve to `realName` exported from `./mod.js`, but does not.

## Expected

`extractDynamicImportNames` needs the same local/original split #1730 introduced for static `import_specifier` handling (`Import.renamedImports` / `RenamedImport`), so `names` carries the local alias `b` and a rename pair records `{ local: 'b', imported: 'a' }` for the resolver to consult.

## Scope note

Deliberately not fixed as part of #1730 — different extraction function (dynamic `import()` destructuring, not static `import` specifiers), or a distinct enough sub-case to warrant its own fix + tests.

### PR DESCRIPTION
**Pull request by carlos-alm**
## Summary

- `const { a: b } = await import(...)` recorded the original exported name (`a`) instead of the local binding actually used at call sites (`b`) — so a subsequent call to `b(...)` never resolved to the imported symbol. The `pair_pattern` branch of `extractDynamicImportNames` (TS/WASM) and `extract_dynamic_import_names` (Rust) preferred tree-sitter's `key` field (the name exported by the target module) over `value` (the local binding) — the same class of bug fixed for static `import { X as Y }` specifiers in #1730.
- Both engines now record the local alias in `names`, with the local → original mapping recorded in `Import.renamedImports` — reusing the exact mechanism #1730/#1823 already threaded through call resolution, barrel tracing, and the incremental path. No resolver changes were needed, only extraction, since that infrastructure already exists and is generic across static/dynamic imports.
- Also handles the combined rename + default-value case (`{ a: b = defaultValue }`) and falls back to the previous key-based behavior for nested patterns (`{ foo: { nested } }`), which have no single local binding to extract.
- As a side effect of fixing the shared Rust `collect_object_pattern_names` helper, CJS `require()` destructuring renames (`const { a: b } = require('./mod')`) are now also correct on the native engine, bringing it in line with WASM's already-correct `extractCjsRequireBinding`.

## Test plan

- `npx vitest run tests/parsers/javascript.test.ts` — 208/208 pass, including 5 new tests for the rename/default/parens+as-cast cases
- `npx vitest run tests/integration/issue-1824-dynamic-import-destructure-rename.test.ts` — new integration test, WASM + native variants, 6/6 pass
- Verified the new integration test actually fails without the fix (rebuilt `dist/` from the pre-fix source, confirmed 2 failures; restored the fix and rebuilt, confirmed all pass)
- `cargo test --lib` — 529 passed, including 3 existing Rust unit tests updated (they previously asserted the buggy behavior as expected output, per the "never document bugs as expected behavior" rule)
- `npm test` — full suite green (3756 passed, 30 skipped, 2 todo)
- `npm run lint` — clean
- `codegraph diff-impact --staged -T` — confirmed blast radius is scoped to `extractDynamicImportNames`/`collectDynamicImport`/`handleDynamicImportCall` and their callers only

Filed #1969 (WASM dynamic-import destructuring misses shorthand-default and rest-element bindings that native already handles) — a distinct, pre-existing parity gap discovered while reading the Rust/WASM object-pattern loops side by side, out of scope for this fix.

Stacked on #1823 (base branch `fix/issue-1823-barrel-re-export-with-rename-export-x-as-y`) — only the extractor/types/test diff is this issue's change.

Fixes #1824

### CODE REVIEW

File: src/extractors/javascript.ts [hunk 1]
```diff
@@ -4204,10 +4222,37 @@ function extractDynamicImportNames(callNode: TreeSitterNode): string[] {
...
-        if (key) names.push(key.text);
+        const value = child.childForFieldName('value');
+        let localNode: TreeSitterNode | undefined;
+        if (
+          value?.type === 'identifier' ||
+          value?.type === 'shorthand_property_identifier_pattern'
+        ) {
+          localNode = value;
+        } else if (value?.type === 'assignment_pattern') {
+          // { imported: local = defaultValue } — the local binding is the
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply src/extractors/javascript.ts:4248]: Fixed — resolved the key to a clean name the same way resolveComputedKeyName already does for object-literal keys elsewhere in this file: string/string_fragment keys get quotes stripped, computed_property_name keys get unwrapped (non-string computed keys like `[Symbol()]` resolve to no name, matching the existing #1764 convention). Also caught a related edge case while in there — the local binding needs to stay tracked in `names` even when the key isn't statically resolvable, which the original structure would have silently dropped entirely. Verified empirically: reproduced the exact bug you described first (`imported: "'foo-bar'"` with quotes literally included), then confirmed the fix for both quoted and computed-string cases, plus the non-string-computed-key fallback. New tests cover all three.

File: crates/codegraph-core/src/extractors/javascript.rs [hunk 2]
```diff
@@ -3124,9 +3148,46 @@ fn collect_object_pattern_names(pattern: &Node, source: &[u8]) -> Vec<String> {
...
+                        // binding is the assignment_pattern's left identifier.
+                        v.child_by_field_name("left")
+                            .filter(|left| left.kind() == "identifier")
+                    }
+                    _ => None,
+                };
+                match (local_node, key) {
+                    (Some(local_node), Some(key)) => {
+                        let local_text = node_text(&local_node, source).to_string();
+                        let key_text = node_text(&key, source).to_string();
```
[the remark this answers is not in the document]
Comment (carlos-alm) [reply crates/codegraph-core/src/extractors/javascript.rs]: Fixed the same way on the native side, mirroring resolve_computed_key_name (already used for object-literal computed keys elsewhere in this file): string/string_fragment keys get quotes stripped via extract_string_fragment, computed_property_name keys get unwrapped, non-string computed keys resolve to no rename pair. Also fixed a related edge case: the local binding now stays tracked in `names` even when the key isn't statically resolvable, rather than being silently dropped along with the unresolvable rename mapping. Added 3 new native unit tests mirroring the TS-side fix. Verified empirically before and after the fix, full suite green (227 files, 3779 tests).

**Comment by carlos-alm** [comment]
@greptileai
